experimental/devices/mcp9808/mcp9808.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package mcp9808
+
+import (
+	"encoding/binary"
+	"errors"
+	"sync"
+
+	"periph.io/x/periph/conn"
+	"periph.io/x/periph/conn/i2c"
+	"periph.io/x/periph/conn/mmr"
+	"periph.io/x/periph/conn/physic"
+)
+
+// Opts holds the configuration options.
+//
+// Slave Address
+//
+// Depending which pins the A0, A1 and A2 pins are connected to will change the
+// slave address. Default configuration is address 0x18 (Ax pins to GND). For a
+// full address table see datasheet.
+type Opts struct {
+	Address    int
+	Resolution resolution
+}
+
+// DefaultOpts is the recommended default options.
+var DefaultOpts = Opts{
+	Address:    0x18,
+	Resolution: Maximum,
+}
+
+// New opens a handle to an mcp9808 sensor.
+func New(bus i2c.Bus, opts *Opts) (*Dev, error) {
+
+	i2cAddress := DefaultOpts.Address
+	if opts.Address != 0 {
+		if opts.Address < 0x18 || opts.Address > 0x1f {
+			return nil, errAddressOutOfRange
+		}
+		i2cAddress = opts.Address
+	}
+
+	dev := &Dev{
+		m: mmr.Dev8{
+			Conn:  &i2c.Dev{Bus: bus, Addr: uint16(i2cAddress)},
+			Order: binary.BigEndian,
+		},
+	}
+
+	if err := dev.setResolution(opts.Resolution); err != nil {
+		return nil, err
+	}
+	if err := dev.enable(); err != nil {
+		return nil, err
+	}
+
+	return dev, nil
+}
+
+// Dev is a handle to the mcp9808 sensor.
+type Dev struct {
+	m mmr.Dev8
+
+	mu       sync.Mutex
+	critical physic.Temperature
+	upper    physic.Temperature
+	lower    physic.Temperature
+	shutdown bool
+}
+
+// Sense reads the current ambient temperature.
+func (d *Dev) Sense() (physic.Temperature, error) {
+	if err := d.enable(); err != nil {
+		return 0, err
+	}
+	tbits, err := d.m.ReadUint16(temperature)
+	if err != nil {
+		return 0, errReadTemperature
+	}
+	// Convert to physic.Temperature 0.0625°C per bit
+	t := physic.Temperature(tbits&0x0FFF) * 62500 * physic.MicroKelvin
+	if tbits&0x1000 > 0 {
+		// Check for bit sign.
+		t = -t
+	}
+	t += physic.ZeroCelsius
+	return t, nil
+}
+
+// SenseWithAlerts reads the ambient temperature and returns an slice of any
```

Comment:
I'm not sure "ambient" is relevant here, the sensor may be strapped on a heater or something. :)